### DISCUSSION TITLE
Can I threshold a grayscale image to [0, 1] range?

### DISCUSSION THREAD

**Discussion by bemoregt** [ImageMagick/ImageMagick#2151] [Help] [answered]
Hi, @alex @mirakui @yoya @rudchenkos @bgK 

Can I threshold a grayscale image to [0, 1] range?

The grayscale image has the [0, 255] ranged color.
and, Thresholded image should have [0, 1] ranged color.

How can I convert that using imagemagick ?

Thanks in advance.
Best,
@bemoregt.

**Comment by fmw42** [comment, marked as answer]
ImageMagick threshold internally to range 0 to 1, but externally it will be 0 to 255. If you tried to get 0 to 1, all you would have would be 0s and 1s in an 8-bit image. To get 0 to 1 will float values between, you would need HDRI install of ImageMagick (default for IM 7), but would need to save to a file format that supports float. That would be either PFM or TIFF with -define quantum:format=floating-point.


`magick grayimage -threshold X% -evaluate multiply 0.000015259021897 result.pfm`

where 0.000015259021897 = 1/quantumrange = 1/65536 for default IM 7.

But why do you need an image in the range 0 to 1?

**Reply by bemoregt** [reply]
Hi, @fmw42 

But why do you need an image in the range 0 to 1?
>> My final goal is to make mask image for Mask R-CNN datasets.
>> those datasets are must have just [0, 1, 2, 3] color values for deep learning training fast speed, may be.

Thanks. I'll try that command.

Best,
@bemoregt.

**Reply by bemoregt** [reply]
Imagemagick's one-lined commands is good to data preprocessing for deep learning datatsets easily in linux or osx's terminal. 

Especially for image data augmentation. etc... that's powerfull !!

Thanks magick~
@bemoregt.

**Reply by fmw42** [reply]
If you want just two values, 0 and 1, why not just use 0 and 255?  What format do you need, 8-bit ints or floats?

If 8-bit values

magick grayimage -threshold X% -evaluate divide 65535 result.png

if you need floats, then

magick grayimage -threshold X% -evaluate divide 65535 result.pfm

Is probably better than -evaluate multiply 0.000015259021897

**Reply by bemoregt** [reply]
I want to get 8bit INT png format, but it's pixel has only [0, 1] color.

Thanks.

**Reply by fmw42** [reply]
try

`magick grayimage -threshold X% -evaluate divide 65535 -type grayscale PNG8:result.png`

**Reply by bemoregt** [reply]
Thanks @fmw42 

I finally did it  with this command:

convert 9_mask.png -threshold 128 -evaluate divide 255 n_9_mask.png

Best,
@bemoregt.

**Reply by bemoregt** [reply]
My final bash script is here:

for fn in *.png; do convert -threshold 128 -evaluate divide 255 "${fn}" n_"${fn}"; done

"In order to increase generalization performance in deep learning, you need to be familiar with ImageMagick."

Thanks. @fmw42 
@bemoregt.
